File: código/avg.py
from math import sqrt, pow

class Mean():
	def __init__(self, vals*):
		if(len(vals) == 1):
			self.values = [float(vals[0])]
		elif(len(vals) == 0):
			self.values = []
		else:
			self.values = map(lambda v: float(v), vals)
	def mean(self):
		return sum(self.values) / self.factor()
	def factor(self):
		return len(self)
class GeometricMean(Mean):
	def __init__(self):
		Mean.__init__(self)
	def mean(self):
		prod = 1
		for v in self.values:
			prod *= v
		return pow(prod, 1 / self.factor())
class QuadraticMean(Mean):
	def __init__(self):
		Mean.__init__(self)
	def mean(self):
		return sqrt(sum(map(lambda v: v ** 2, self.values)) / self.factor())
class HarmonicMean(Mean):
	def __init__(self):
		Mean.__init__(self)
	def mean(self):
		return self.factor() / sum(map(lambda v: 1 / v, self.values))

File: código/avg.js
class Mean extends Array{
	constructor(...vals){
		const c = this.constructor;
		if(vals.length === 1)
			super(Number(vals));
		else if(vals.length === 0)
			super();
		else
			super(...c.filter(vals));
	}
	push(...v){
		super.push(...c.filter(v));
	}
	static filter(arr){
		return arr.map(v => parseFloat(v)).filter(isFinite);
	}
	toNumber(){
		return this.mean;
	}
	valueOf(){
		return this.mean;
	}
	get mean(){
		return this.reduce((a, b) => a + b, 0) / this.factor;
	}
	get factor(){
		return this.length;
	}
}
class GeometricMean extends Mean{
	get mean(){
		return Math.pow(this.reduce((a, b) => a * b, 1), 1 / this.factor);
	}
}
class QuadraticMean extends Mean{
	get mean(){
		return Math.sqrt(this.map(v => v ** 2).reduce((a, b) => a + b, 0) / this.factor);
	}
}
class HarmonicMean extends Mean{
	get mean(){
		return this.factor / (this.map(v => v ** -1).reduce((a, b) => a + b, 0));
	}
}

module.exports = Mean;
Mean.GeoMean = GeometricMean;
Mean.QuadMean = QuadraticMean;
Mean.HarmMean = HarmonicMean;